Inference with WaveformArray

cdcapano opened the pull request:
These patches update `io/inference_hdf.py` so that it loads samples as instances of `WaveformArray`. This makes it easy to plot/summarize functions of the variable args, along with derived parameters of the variable, such as chirp mass. It also makes use of the new parameters module, so automatically get labels of unrecognized parameters. This change to `inference_hdf` necessitates updates to the programs that load the samples. Since these programs make use of some common command line options --- namely, `input-file, `parameters` (which replaces the old `variable-args`), `thin-start`, and `thin-interval` --- I also created an `option_utils` module in inference that stores and loads these options.

As before, you need only provide an input-file and an output-file to plot the basic variable-args in the file. For example,
`pycbc_inference_plot_posterior --input-file cbc_example-n1e4_sr2048.hdf --output-file corner_plot-default.png`
Yields:
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/wfarray_in_inference/corner_plot-default.png

However, now you can specify derived parameters from WaveformArray, as well as functions of the variable args and these derived paramters. For example, the following plots the primary and secondary masses (defined as m_pr >= m_sc), as well as the coalesence time - a GPS time in miliseconds:
`pycbc_inference_plot_posterior --input-file cbc_example-n1e4_sr2048.hdf --output-file corner_plot-mp_ms_tc.png --parameters '(tc-1126259462)*1000.:$t_c - 1126259462$ (ms)' m_p m_s`
This yields:
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/wfarray_in_inference/corner_plot-mp_ms_tc.png
Note that custom labels are now specified by placing a colon after a parameter with the label, rather than a separate label option. This way, if you only want to customize one label (as in the above), you do not need to re-write all of the labels on the command line.

Here is an example of plotting chirp mass:
`pycbc_inference_plot_posterior --input-file cbc_example-n1e4_sr2048.hdf --output-file corner_plot-mchirp.png --parameters mchirp`
which yields:
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/wfarray_in_inference/corner_plot-mchirp.png

An example with `table_summary`:
`pycbc_inference_table_summary --output-file summary_table-tc_mp_ms_mchirp_eta_mtotal_q.html --input-file cbc_example-n1e4_sr2048.hdf --verbose --parameters '(tc-1126259462)*1000.:$t_c - 1126259462$ (ms)' m_p m_s mchirp eta mtotal q`
Yields:
https://www.atlas.aei.uni-hannover.de/~cdcapano/public_scratch/wfarray_in_inference/summary_table-tc_mp_ms_mchirp_eta_mtotal_q.html
There is still an issue with the latex to html conversion, as not all of the symbols are being converted, but that will be addressed in a future pull request.

I've also updated `plot_acf` `plot_acl`, and `plot_samples`.  I left `plot_prior` and `plot_acceptance_rate` alone since they do not use `read_samples(_from_walker)`.

cmbiwer commented:
```
Traceback (most recent call last):
  File "/home/cbiwer/opt/pycbc-test/bin/pycbc_inference", line 4, in <module>
    __import__('pkg_resources').run_script('PyCBC===eae1e0', 'pycbc_inference')
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/pkg_resources/__init__.py", line 719, in run_script
    self.require(requires)[0].run_script(script_name, ns)
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/pkg_resources/__init__.py", line 1504, in run_script
    exec(code, namespace, namespace)
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/PyCBC-eae1e0-py2.6.egg/EGG-INFO/scripts/pycbc_inference", line 153, in <module>
    with InferenceFile(opts.output_file, "w") as fp:
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/PyCBC-eae1e0-py2.6.egg/pycbc/io/inference_hdf.py", line 85, in __init__
    if self.variable_args[0] in self:
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/PyCBC-eae1e0-py2.6.egg/pycbc/io/inference_hdf.py", line 110, in variable_args
    return self.attrs["variable_args"]
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/usr1/cbiwer/pip-build-bhqd0P/h5py/h5py/_objects.c:2687)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/usr1/cbiwer/pip-build-bhqd0P/h5py/h5py/_objects.c:2645)
  File "/home/cbiwer/opt/pycbc-test/lib/python2.6/site-packages/h5py/_hl/attrs.py", line 58, in __getitem__
    attr = h5a.open(self._id, self._e(name))
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (/usr1/cbiwer/pip-build-bhqd0P/h5py/h5py/_objects.c:2687)
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (/usr1/cbiwer/pip-build-bhqd0P/h5py/h5py/_objects.c:2645)
  File "h5py/h5a.pyx", line 77, in h5py.h5a.open (/usr1/cbiwer/pip-build-bhqd0P/h5py/h5py/h5a.c:2175)
KeyError: "Can't open attribute (Can't locate attribute: 'variable_args')"
```

This PR breaks ``pycbc_inference``. Error is in trying to initialize. There is no ``InferenceFile.variable_args``.

Also this is more of a style thing but I see you renamed ``variable_args`` in certain functions to ``parameters``. I think it best to keep to uniform if possible, they appear to be logically the same things (ie. a list of str of parameter names).

cmbiwer commented on pycbc/io/inference_hdf.py:
```diff
@@ -121,84 +142,73 @@ def acl(self):
-
-        return self[variable_arg]["walker%d"%nwalker][thin_start::thin_interval]
+        # figure out the size of the output array to create
+        #n_per_walker = 1 + \
```
You got some lines commented out here. Is there a reason to keep them around?

cdcapano commented:
Sorry about that. I just pushed a patch to fix the bug when initializing a new file. I also removed the commented out code.

I changed the name of the command line argument from `variable-args` to `parameters` because I wanted to distinguish the two. The `variable-args` are the specific parameters that were varied in the MCMC. However, the `parameters` given to the plotting/summary codes can be functions of, or derived from, the `variable-args`. For instance, you can tell `plot_posterior` to plot `mchirp` even though mchirp was not a variable-arg in the MCMC; `mass1` and `mass2` were.

My thought was changing the command-line name would avoid confusion, but I realize this might be a subtle distinction. If you think it's actually more confusing, I'll change the command-line option back to `variable-args`.

cmbiwer commented:
Fair enough about ``--parameters``.

This PR also breaks the workflow generator because it renames``-variable-args`` but that's easy enough to fix I will merge this and make another PR.
